Add Phiro-Flavor

sjaindl opened the pull request.

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	private void removePhiroPrograms() {
+		for (String project : PhiroMainMenuActivity.phiroProjects) {
```
there should never be any programs existing in this state (before a test is started, when setup is called) why is this necessary?

sjaindl replied:
In the case of Phiro there may be programs:
There are programs (placed in .zip-Files in the assets-folder) that are extracted when MyProjectsActivity is opened. So this is necessary to ensure that no programs exist in the first place.. It could be done in in teardown, alternatively

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+		onView(withText(R.string.main_menu_programs))
+				.perform(click());
+
+		ViewInteraction viewInteraction = onView(withId(android.R.id.list));
```
please dont call this "viewInteraction" because pretty much everything in espresso is either a view / data Interaction and this does not help the reader in any way.

84n4n4 commented on catroid/src/androidTest/java/org.catrobat.catroid/uiespresso/util/matchers/AdapterViewMatchers.java:
```diff
@@ -0,0 +1,84 @@
+
+			@Override
+			public boolean matchesSafely(View view) {
+				return matcher.matches(view) && currentIndex++ == index;
```
what is the "++" good for? (also, currentIndex is first compared with index and incemented AFTERWARDS, so its not doing anything here (according to java guru cheidenreich))

sjaindl replied:
I'm not quite sure about when exactly matchesSafely is called (you are the Espresso Guru :)), but it's a member variable. So I think it might be to iterate through a list, for example. In the first pass it's 0, then 1 etc. until it matches..
 (I've taken and adapted that matcher from https://stackoverflow.com/questions/29378552/in-espresso-how-to-choose-one-the-view-who-have-same-id-to-avoid-ambiguousviewm)
So, I'm not quite sure, but I'll try without incrementing whether the test passes

84n4n4 replied:
but you do have an adapter view, and should use onData in the first place, instead of writing a matcher that does something that espresso can do for you out of the box.
the stackoverflow answer has some stuff not in a adapterView and uses this approach because of this.

write a matcher that ensures you got the right list on your hands similar to this one here:
https://github.com/Catrobat/Catroid/blob/develop/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/BrickCategoryListMatchers.java

and then use it like this:

onData(allOf(is(instanceOf(String.class)), is(UiTestUtils.getResourcesString(categoryStringResource))))
					.inAdapterView(BrickCategoryListMatchers.isBrickCategoryView())
					.check(matches(isDisplayed()));

found in https://github.com/Catrobat/Catroid/blob/develop/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/BrickValueParameterTest.java

or even add a wrapper that calls above chunk of code and gives you a data interaction like this:
public static BrickDataInteractionWrapper onBrickAtPosition(int position) {
		return new BrickDataInteractionWrapper(onData(instanceOf(Brick.class))
				.inAdapterView(ScriptListMatchers.isScriptListView())
				.atPosition(position));
	}

https://github.com/Catrobat/Catroid/blob/develop/catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/content/brick/utils/BrickDataInteractionWrapper.java

and when using espresso functionality instead of your own matcher that does "atPosition", "inAdapterView", "onChildView", etc...


and then

sjaindl replied:
Thanks for your suggestions, they will be useful - I'm currently rewriting it to use onData(..)

84n4n4 commented on catroid/src/androidTest/java/org.catrobat.catroid/uiespresso/util/matchers/AdapterViewMatchers.java:
```diff
@@ -0,0 +1,84 @@
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+
+public final class AdapterViewMatchers {
```
this name is so broad that it does not add any information or hint on what this is exactly good for.

84n4n4 commented on catroid/src/androidTest/java/org.catrobat.catroid/uiespresso/util/matchers/AdapterViewMatchers.java:
```diff
@@ -0,0 +1,84 @@
+		throw new AssertionError();
+	}
+
+	public static Matcher<View> withAdaptedDataAt(final String expectedContent, final int listPosition) {
```
calling a view matcher withAdapted**Data**At is not a good idea, since in espresso we got view- and data- Matchers/Actions/etc, and those are different things

84n4n4 commented on catroid/src/androidTest/java/org.catrobat.catroid/uiespresso/util/matchers/AdapterViewMatchers.java:
```diff
@@ -0,0 +1,84 @@
+
+				Adapter adapter = ((AdapterView) view).getAdapter();
+				String listItem;
+				if (adapter instanceof ProjectListAdapter) {
```
i have the feeling that all of this shouldnt be done with a custom view matcher but with onData functionality of espresso in the first place.

sjaindl replied:
Yes, it's now done with onData and custom view matchers/wrappers

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+		checkBrickInProject();
+	}
+
+	private void addBrickToProject(String projectName) {
```
what is this good for in the first place?

sjaindl replied:
This tests the following scenario:
1. An existing Phiro program is changed
2. User navigates back to MainMenu
3. User navigates to ProjectList
4. The changed Phiro program should be in the same state (not overridden)

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	@Test
+	public void phiroProgramsUnchangedWhenModifiedTest() {
```
write a separate test class that sets up a project, and then checks if it is properly displayed, and add/remove bricks there.

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+	private void addBrickToProject(String projectName) {
+		try {
+			ProjectManager.getInstance().loadProject(projectName, InstrumentationRegistry.getTargetContext());
+		} catch (LoadingProjectException e) {
```
why catch exceptions in a test without having it fail?!?

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+		project.getDefaultScene().getSpriteList().get(0).getScript(0).addBrick(new NoteBrick("I was added!"));
+	}
+
+	private void checkBrickInProject() {
```
please rename this to provide some info on what this is actually doing

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+		Project project = ProjectManager.getInstance().getCurrentProject();
+		Script script = project.getDefaultScene().getSpriteList().get(0).getScript(0);
+		Brick brick = script.getBrick(script.getBrickList().size() - 1);
+		assertTrue("Brick has been deleted!", brick instanceof NoteBrick);
```
what is that assertion message supposed to tell me?

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroScriptActivityTest.java:
```diff
@@ -0,0 +1,119 @@
+	private void preparePhiroSettings() {
+		SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(InstrumentationRegistry.getTargetContext());
+
+		SharedPreferences.Editor editor = sharedPreferences.edit();
```
when changing shared preferences in a test you should reset them to what they were before the testrun in a teardown method.

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroScriptActivityTest.java:
```diff
@@ -0,0 +1,119 @@
+				.check(matches(withAdaptedDataAt("phiro_front_left_sensor", 0)));
+	}
+
+	public static Project createProject(String projectName) {
```
is there a reason for "public static Project" signature? if not so please make private void

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+		Integer position = 0;
+		for (String program : PhiroMainMenuActivity.phiroProjects) {
+			viewInteraction
+					.check(matches(withAdaptedDataAt(program, ++position)));
```
please add a separate line for incrementing the position.

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	@Test
+	public void phiroProgramsRestoredWhenDeletedTest() {
```
is this test necessary?
shouldnt it be enough to check that projects are generated on program start?

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroScriptActivityTest.java:
```diff
@@ -0,0 +1,119 @@
+		onView(withId(R.id.formula_editor_keyboard_sensors))
+				.perform(click());
+
+		onView(withIndex(withId(android.R.id.list), 1))
```
please use onData

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,172 @@
+		for (String setting : allSettings) {
+			initialSettings.put(setting, sharedPreferences.getBoolean(setting, false));
+		}
+		initialSettings.put(SETTINGS_SHOW_PHIRO_BRICKS, sharedPreferences.getBoolean(SETTINGS_SHOW_PHIRO_BRICKS, false));
```
please include SETTINGS_SHOW_PHIRO_BRICKS in allSettings to make this work as intended (and also to have the teardown work as expected, also deactivating SETTINGS_SHOW_PHIRO_BRICKS again!!!)

sjaindl replied:
That's intentionally. I want to test whether the setting is initially activated - So I don't want it to be set to true in setAllSettingsTo(true) .. but true, it should be restored in teardown, nevertheless. I'll add that!

84n4n4 replied:
if the test is about checking out the default / initial state, why touch shared preferences in the first place? 
also this should be verified via UI, since you have to give a fallback when asking sharedPreferences, and if you do sharedPreferences.getBoolean(SETTINGS_SHOW_PHIRO_BRICKS, false) in your test, doesnt mean that someone might have used sharedPreferences.getBoolean(SETTINGS_SHOW_PHIRO_BRICKS, true) in production code, (would make a difference if run for the first time and key isnt there yet)

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,172 @@
+
+	@Test
+	public void basicSettingsTest() {
+		checkPreference(R.string.preference_title_enable_arduino_bricks, SETTINGS_SHOW_ARDUINO_BRICKS);
```
this is already tested in SettingsActivityTest, is there any point in duplicating this test?

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/ui/MainMenuActivity.java:
```diff
@@ -375,16 +387,10 @@ public void handleWebButton(View view) {
-		} else {
-			startWebViewActivity(Constants.BASE_URL_HTTPS);
-		}
+		String url = Constants.FLAVORED_BASE_URL_HTTPS;
```
isnt this going to break the webview login if above code is removed?

sjaindl replied:
I've moved that code to a Utils function, as this is used in other places, too.
--> This is now done in: Utils.addUsernameAndTokenInfoToUrl(..)

84n4n4 commented:
btw, whats up with the hundreds of filepath changes? 

Catroid/catroid/src/androidTest/java/org/catrobat/catroid/... 
  to
Catroid/catroid/src/androidTest/java/org.catrobat.catroid/...

84n4n4 commented:
had a further look into the files that had their paths changed, was this intentional:
catroid/src/**androidTest**/assets/Air_fight_0.5.catrobat 
→ 
catroid/src/**androidTestCatroid**/assets/Air_fight_0.5.catrobat

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,172 @@
+		onData(PreferenceMatchers.withTitle(R.string.preference_title_enable_phiro_bricks)).perform(click());
+		checkPreference(R.string.preference_title_enable_phiro_bricks, SETTINGS_SHOW_PHIRO_BRICKS);
+
+		String phiroLink = InstrumentationRegistry.getTargetContext().getString(R.string.phiro_preference_link);
```
theres a 
public static String getResourcesString(int stringId) 

in espresso UiTestUtils that does this.

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/ui/fragment/FragmentFactory.java:
```diff
@@ -0,0 +1,64 @@
+import android.app.Fragment;
+import android.content.Context;
+
+final class FragmentFactory {
```
very general name if this is only meant to provide BrickCategoryFragment and FormulaEditorCategoryListFragment.
although its named factory, but its not really analog to a GOF factory pattern.

wouldnt it be nicer to use static factory methods in those fragments, that provide the right implementation when called (http://www.androiddesignpatterns.com/2012/05/using-newinstance-to-instantiate.html)

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/ui/fragment/FragmentFactory.java:
```diff
@@ -0,0 +1,64 @@
+	}
+
+	private static Fragment createFragment(String className, Context context) {
+		String classNameWithPackage = context.getPackageName() + "." + className;
```
i dont know a lot about the whole flavoring thing, but is this really supposed to be done like this? (like building the full class and package as string and the loading it that way)

84n4n4 replied:
ok, i read up on android flavoring, and all of this code, including the factory and deciding on runtime what implementation to initiate is not how flavoring is meant to be used.
please change structure.
https://stackoverflow.com/questions/23698863/build-flavors-for-different-version-of-same-class

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/ui/fragment/FragmentFactory.java:
```diff
@@ -0,0 +1,64 @@
+	private static Fragment createFragment(String className, Context context) {
+		String classNameWithPackage = context.getPackageName() + "." + className;
+
+		if (!classNameExists(classNameWithPackage, context)) {
```
i see this rather problematic, this could make bugs really hard to find, also this is something that shouldnt happen if everything is correct in the first place, right?
if a classname / packagename would be malformed or not existing, this would still make sure that everything compiles and runs, but is this really what we want? (id rather see this break on runtime in any test, then accidentially rolling this out some time and having to find out that the wrong fragment is displayed)

84n4n4 commented:
and same goes for all tests, the common test base stays in androidTest and the flavor specific ones go into their flavor androidTestPhiro and androidTestCatroid (while androidTestCatroid would be empty at this point)
https://stackoverflow.com/questions/22373655/test-package-for-different-flavors-in-android-studio
http://tools.android.com/tech-docs/new-build-system/user-guide#TOC-Testing1

cheidenreich commented on catroid/src/main/java/org/catrobat/catroid/content/bricks/ConcurrentFormulaHashMap.java:
```diff
@@ -30,6 +32,7 @@
 
 	private static final long serialVersionUID = 9030965461744658052L;
 
+	@TargetApi(24)
```
why is this necessary?

cheidenreich commented on catroid/src/main/java/org/catrobat/catroid/scratchconverter/ScratchConversionManager.java:
```diff
@@ -150,6 +150,13 @@ private void closeAllActivities() {
 					Log.e(TAG, e.getMessage());
 				}
+			} else if (BuildConfig.PHIRO_CODE) {
+				try {
```
duplicate code, you can just set the string as a variable

cheidenreich commented on catroid/src/main/java/org/catrobat/catroid/ui/SettingsActivity.java:
```diff
@@ -553,7 +579,7 @@ public static boolean isNfcSharedPreferenceEnabled(Context context) {
 		SharedPreferences.Editor editor = getSharedPreferences(context).edit();
 		editor.putBoolean(SETTINGS_SHOW_NFC_BRICKS, value);
-		editor.commit();
+		editor.apply();
```
`putBoolean()` returns the editor again, so you can just append the `.apply()` above

cheidenreich commented on catroid/src/main/java/org/catrobat/catroid/ui/fragment/BrickCategoryFragment.java:
```diff
@@ -71,6 +75,61 @@ public void setBrickAdapter(BrickAdapter brickAdapter) {
+		initBrickCategories();
+	}
+
+	private void initBrickCategories() {
```
is this used somewhere else? otherwise it's unnecessary to put in a separate method

sjaindl commented:
I've refactored the flavoring of fragments, should be fine now. 
Adaptations from the main fragment versions only need to be done in the flavor files (if there are no adaptations, the files are empty). The main versions are abstract so that they can't be instantiated.

cheidenreich commented on catroid/src/androidTestCreateatschool/AndroidManifest.xml:
```diff
@@ -0,0 +1,53 @@
+  ~ You should have received a copy of the GNU Affero General Public License
+  ~ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  -->
+<manifest package="org.catrobat.catroid.phiro.test"
```
seems like this file is not meant to be in the folder `catroid/src/androidTestCreateatschool` if its for the test package `org.catrobat.catroid.phiro.test` or am i missing something?

sjaindl replied:
good catch. This Manifest actually isn't necessary anymore .. I've removed it

cheidenreich commented on catroid/src/main/java/org/catrobat/catroid/ui/fragment/MainFormulaEditorCategoryListFragment.java:
```diff
@@ -606,28 +562,44 @@ private void closeCategoryListFragment() {
 		private Map<Integer, String> headerOrderMap = new TreeMap<>();
 
-		public void addHeader(String text) {
+		boolean addCategoryIfDoesNotExist(String header, int[] items, int[] parameters) {
```
why do you have a `boolean` return value if you never check it?

sjaindl replied:
The return value is checked multiple times in MainFormulaEditorCategoryListFragment and FormulaEditorCategoryListFragment of the Phiro flavor

cheidenreich commented on catroid/src/phiro/java/org/catrobat/catroid/ui/fragment/FormulaEditorCategoryListFragment.java:
```diff
@@ -0,0 +1,54 @@
+	}
+
+	private void initCategories() {
+		initSensors();
```
this method is not necessary if the only thing it does is one method call

sjaindl replied:
I've refactored that now so that the code is even more maintainable using the template method pattern. 

This means MainFormulaEditorCategoryListFragment knows the "recipe" of which categories to setup. The "recipe" is initCategories:

	private void initCategories() {
		initObjects();
		initFunctions();
		initLogic();
		initSensors();
	}
These init-methods are protected, so that the subclass can override it to setup the categories with help of polymorphism. 

So, all that needs to be done now in the "Phiro-"FormulaEditorCategoryListFragment is override the method initSensors() and call super.initSensors()

Even much shorter and better maintainable.

cheidenreich commented:
@thmq @84n4n4 any feedback for the latest commit? https://github.com/Catrobat/Catroid/pull/2416/commits/46da464df7220496e2fa428674acdb5402179939

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/utils/FormulaEditorCategoryListDataInteractionWrapper.java:
```diff
@@ -0,0 +1,48 @@
+		super(dataInteraction);
+	}
+
+	public static FormulaEditorCategoryListDataInteractionWrapper onFormulaEditorCategoryItemWithNameAtPosition(String formulaCategoryItem, int position) {
```
i want to have the dataInteraction and viewInteraction wrappers to handle and feel the same as the erspresso internal ones, please have a look at BrickDataInteractionWrapper for example.

so to have every method having only one parameter, i dont want to combine the component and position in one method.
eg.

`public static FormulaEditorCategoryListDataInteractionWrapper onFormulaEditorCategory(String category) {
return new FormulaEditorCategoryListDataInteractionWrapper(onData(allOf(instanceOf(String.class), withFormulaEditorCategoryItemName(formulaCategoryItem)))
inAdapterView(isFormulaEditorCategoryListView());
}`

which then can be used as follows:
instead of 
`onFormulaEditorCategoryItemWithNameAtPosition("someCategory", 5)`
that would be 
`onFormulaEditorCategoryItemWithName("someCategory").atPosition(5)`

but then again, depending on if this is used in multiple test classes or not, it could be even put there as a function instead of separate class...

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/utils/ScriptListDataInteractionWrapper.java:
```diff
@@ -0,0 +1,48 @@
+		super(dataInteraction);
+	}
+
+	public static ScriptListDataInteractionWrapper onBrickCategoryWithNameAtPosition(String brickCategory, int position) {
```
same here as for FormulaEditorCategoryListDataInteractionWrapper

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,153 @@
+		checkAddedNoteBrickIsStillInFirstScript();
+	}
+
+	private void openProjectList() {
```
since all tests start with a call to "openProjectList", refactor the test to open the fragment or activity you actually want to test instead of the main menu activity and delete this function.

sjaindl replied:
There was a reason for doing so.
In MainMenuActivity (Phiro-flavor file) was some initialization logic going on. The Phiro programs have been loaded there when the app is started.

I've moved that code to the flavored ProjectListFragment, where the phiro programs are loaded async now. The downside is a bit of delay until programs are available.

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,153 @@
+				.perform(click());
+	}
+
+	private void addNoteBrickToFirstScript(String projectName) {
```
is there a reason to do this via the UI instead of just modifying the project directly?

sjaindl replied:
No there's no reason to this via the UI. This is just to test whether programs are overwritten.

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/utils/FormulaEditorCategoryListDataInteractionWrapper.java:
```diff
@@ -0,0 +1,48 @@
+	}
+
+	public static FormulaEditorCategoryListDataInteractionWrapper onFormulaEditorCategoryItemWithNameAtPosition(String formulaCategoryItem, int position) {
+		return new FormulaEditorCategoryListDataInteractionWrapper(onData(allOf(instanceOf(String.class), withFormulaEditorCategoryItemName(formulaCategoryItem)))
```
also to be noted, if this whole class will contain only this one static function, then theres no point in having it in a wrapper class in the first place.
the wrappers are intended to provide a way to perform more complicated actions on a certain object while keeping the espresso "look and feel". please have a look at:
https://confluence.catrob.at/display/CAT/Espresso+testing+-+Using+wrappers+instead+of+utility+functions

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroScriptActivityTest.java:
```diff
@@ -0,0 +1,126 @@
+
+		String firstPhiroSensorString = UiTestUtils.getResourcesString(R.string.formula_editor_phiro_sensor_front_left);
+
+		onFormulaEditorCategoryItemWithNameAtPosition(firstPhiroSensorString, 0)
```
ok, regarding what ive wrote further up in the wrapper function. now seeing how this is actually applied, it would be even nicer and way more readable if one would just do:
`onFormulaEditorCategory().atPosition(0).check(matches(withText("someCategoryName")))`

sjaindl replied:
OK, this is possible now for both, FE and Brick categories:
onFormulaEditorCategory()
				.atPosition(0)
				.check(matches(withText(firstPhiroSensorString)));
..and:
onBrickCategory()
				.atPosition(0)
				.check(matches(withText(expectedCategoryString)));

I've left them in wrapper functions since I had to check for the childView(..) of the respective LinearLayout, too. I think therefore it's worth to put it inside a wrapper.

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,175 @@
+		Intents.release();
+	}
+
+	private void openSettings() {
```
use BaseActivityInstrumentationRule<SettingsActivity> to open directly the activity under test.
DO NOT use the ui to navigate to the activity that you want to test. this practice was one of the main reasons the old ui test setup was slow as fuck and flaky.

please have a look at:
https://confluence.catrob.at/display/CAT/Espresso+testing+-+How+to
theres a whole paragraph about this "Open the activity you want to test directly"

sjaindl replied:
There was the same reason as above. I've moved the initialization code to the flavored version of SettingsActivity now.

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/ui/BaseActivity.java:
```diff
@@ -174,6 +178,15 @@ private boolean isRecoveredFromCrash() {
 		return PreferenceManager.getDefaultSharedPreferences(this).getBoolean(RECOVERED_FROM_CRASH, false);
 	}
 
+	private Intent createSettingsIntent() {
```
since this change in funcitonality is depending solely on the flavor, why not cover this also by moving the SettingsActivity into the flavor folders?

sjaindl replied:
Yes, that's one thing I've planned to refactor, too.
OK, then I'll refactor it directly in this branch.

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/ui/WebViewActivity.java:
```diff
@@ -194,6 +194,13 @@ public void onPageStarted(WebView view, String urlClient, Bitmap favicon) {
 					} catch (ClassNotFoundException e) {
 						Log.e(TAG, e.getMessage());
 					}
+				} else if (BuildConfig.PHIRO_CODE) {
```
same here, why not use flavored classes to do this instead of Class.forName based on a build flag?

sjaindl replied:
yes, same as above, this won't be necessary anymore, after all.

84n4n4 commented on catroid/src/main/java/org/catrobat/catroid/utils/TrackingUtil.java:
```diff
@@ -61,7 +61,7 @@ private TrackingUtil() {
 	}
 
 	public static void trackCreateProgram(String projectName, Boolean landscapeMode, boolean exampleProgram) {
```
just looking at the div below, i see the very same if clauses about 50 times.
why not implement a trackerClass that is then instantiated ONCE based on the build flag at startup, either as activeTracker, or as "blackHole" or whatever thats just swallowing tracking calls?

sjaindl replied:
An even better approach might be to use a protection proxy.
I could implement a TrackingInvocationHandler that proxies the TrackingUtil class.
By doing that, all that needs to be done is to check:
if (BuildConfig.DATA_TRACKING) {
	return method.invoke(trackingUtil, method, args);
}
once in the invocation handlers invoke-method. So I don't even need a "blackHole"-Tracker.

sjaindl commented:
OK, next round.
I've refactored the other activities, too. As well as tests (I had to move some initialization code - see comments to outdated diffs) and introduced some design patterns where useful (proxy pattern for tracking, template method for FE-Category initialization).

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/utils/FormulaEditorCategoryListDataInteractionWrapper.java:
```diff
@@ -0,0 +1,47 @@
+		super(dataInteraction);
+	}
+
+	public static DataInteraction onFormulaEditorCategory() {
```
i cant see my comments anymore from before, but this seems to be unchanged. theres no point in having a wrapper if it doesnt provide additional functionality. this is merely a utility function class, that provides one method to instantiate it:
onFormulaEditorCategory()

please at least rename it to FormulaEditorCategoryListDataInteractionUtils, since this is NOT a wrapper

84n4n4 replied:
bump.
i did not meant just renaming this, but leaving it extending the DataInteractionWrapper.

please take
FormulaEditorCategoryListDataInteractionUtils.java
ProjectListDataInteractionUtils.java
ScriptListDataInteractionUtils.java

and move the one contained functions somewhere meaningful, in this case, since each of them is only used in single test classes, put them in the testclasses.

so in general, if a utility function is used in - 
- one test class, put it there
- multiple test classes in the same package, make a utility function class in this package, make it package private
- multiple test classes spanning multiple distinct packages, make a global utility file with a meaningful name in the utils packag.

sjaindl replied:
OK, I've moved them to the respective test classes.

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/utils/ProjectListDataInteractionWrapper.java:
```diff
@@ -0,0 +1,49 @@
+		super(dataInteraction);
+	}
+
+	public static ProjectListDataInteractionWrapper onProjectWithName(String projectName) {
```
same here as with FormulaEditorCategoryListDataInteractionWrapper

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/ui/activity/utils/ScriptListDataInteractionWrapper.java:
```diff
@@ -0,0 +1,47 @@
+		super(dataInteraction);
+	}
+
+	public static DataInteraction onBrickCategory() {
```
and here aswell...

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/FormulaEditorCategoryListMatchers.java:
```diff
@@ -0,0 +1,68 @@
+		return new TypeSafeMatcher<View>() {
+			@Override
+			protected boolean matchesSafely(View item) {
+				return item.getId() == R.id.fragment_formula_editor_list_item;
```
why not user withId(R.id.fragment_formula_editor_list_item) since this is only used once?

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/FormulaEditorCategoryListMatchers.java:
```diff
@@ -0,0 +1,68 @@
+		throw new AssertionError();
+	}
+
+	public static Matcher<View> isFormulaEditorCategoryListView() {
```
this is doing exactly the same as util.matchers.BrickCategoryListMatchers

84n4n4 commented on catroid/src/androidTest/java/org/catrobat/catroid/uiespresso/util/matchers/ProjectListMatchers.java:
```diff
@@ -0,0 +1,76 @@
+
+			@Override
+			public boolean matchesSafely(final ProjectData actualObject) {
+				return expectedName.matches(actualObject.projectName);
```
please rename actualObject to actualProjectData

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,126 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.uiespresso.ui.activity;
```
can you put the phiro tests in a separate package please

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroMyProjectsActivityTest.java:
```diff
@@ -0,0 +1,126 @@
+		baseActivityTestRule.launchActivity(null);
+	}
+
+	private void addNoteBrickToFirstScript(String projectName) {
```
please remove everything from this function besides:
 +		project.getDefaultScene().getSpriteList().get(0).getScript(0).addBrick(new NoteBrick("I was added!")); 

since unnecessary switching around activities leads to problems with espresso tests when running in bulk. also, i think that testing if the project is really opened when clicked, should be tested in a seperate test dealing with basic functionality of the projectManager

sjaindl replied:
I can't directly say that, because there's no project loaded at that stage. So, I had to set the project first..

84n4n4 replied:
ok

84n4n4 replied:
but in general ive got a bad feeling with this test, since my experience has shown that manually killing / relaunching / and especially switching inbetween activities leads to really weird test failures when the test is run in conjunction with other tests. (did that a few times already, was a painful learning experience)

ill try to have this run on jenkins to see if this causes other tests to fail.

84n4n4 replied:
ok, so heres the run on your repository:
https://jenkins.catrob.at/view/LeeroyCatroid/job/CatroidSinglePackageEmulatorTest/19319/

and here the one on the feature/Nolb
https://jenkins.catrob.at/view/LeeroyCatroid/job/CatroidSinglePackageEmulatorTest/19320/

since most of the tests are in another package, they arent run here in the first place...
weve got to think up a proper way on how to deal with the flavored tests i guess...

84n4n4 replied:
both look similar, but also the phiroActivity tests arent run there...

sjaindl replied:
Doesn't sound so well. What was your solution to those cases where you had to restart/switch activities?
OK, so the same tests are failing.. are they failing in develop, too? Since they shouldn't have something to do with feature/NOLB in the first place. 
Ooh yeah we have to think about that.. maybe tests could be build and executed on a per flavor basis on Jenkins?

sjaindl replied:
Hmm, at least in feature/nolb there are really no flavored Espresso tests yet

84n4n4 replied:
yeah, the ones above are failing in develop aswell, there is actually a bug causing them to fail.

anyways, i dont really have a solution for the activity switching issue. with this tests here weve got a change that its going to be fine since you start / finish / restart the activity under test using the test rule.

the problems i encountered up to this point was not that problematic tests were failing, but they made other tests fail. and to make things more dificult, not the next ones, but there was some constellation that the next ones running main menu activity were failing.
although ive read about other people having similar problems, its worse in our project due to the huge number of activities, and us messing with the back stack with certain activities, and also the sometimes kind of crappy code in some of the activity lifecylce related methods (onCreate, onDestroy, etc...)
so in the end i just started removing tests from the testsuites to see when the other test was working again and could narrow it down.

so all in all, i dont really have a solution for this problem.....
so its just something i try to avoid in the first place...

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroScriptActivityTest.java:
```diff
@@ -0,0 +1,130 @@
+		editor.apply();
+	}
+
+	private static void createProject(String projectName) {
```
theres a createProjectAndGetStartScript in content.brick.utils.BrickTestUtils that could be used instead:
eg:
BrickTestUtils.createProjectAndGetStartScript(projectname).addBrick(new SetXBrick());

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,151 @@
+	}
+
+	@Test
+	public void phiroIsInitiallyActivatedTest() {
```
in lite 86 in setup all settings are set to false: setAllSettingsTo(false)

so this test does merely verify the setup method of itself

sjaindl replied:
It does, in setup all settings are set to false. This checks whether it has been set to true again when starting MyProjectsActivity!

84n4n4 replied:
ok

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,151 @@
+	@Test
+	public void phiroIsFirstVisibleSettingTest() {
+		onData(PreferenceMatchers.withTitle(R.string.preference_title_enable_phiro_bricks))
+				.atPosition(0)
```
why is the atPosition necessary here? there shouldnt be more than one preferences with the same title, if so that should fail rightfully

sjaindl replied:
This is necessary here because I want to explicitly check whether this preference is the first in the list. Without .atPosition(0) the test would pass, but shouldn't!

84n4n4 replied:
ok

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,151 @@
+		sharedPreferencesEditor.commit();
+	}
+
+	private void checkPreference(int displayedTitleResourceString, String sharedPreferenceTag) {
```
sinc this function is idential with the one in SettingsActivityTest, please move it to a utility class

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,151 @@
+		initialSettings.clear();
+	}
+
+	private void setAllSettingsTo(boolean value) {
```
sinc this function is idential with the one in SettingsActivityTest, please move it to a utility class

84n4n4 commented on catroid/src/androidTestPhiro/java/org/catrobat/catroid/uiespresso/ui/activity/PhiroSettingsActivityTest.java:
```diff
@@ -0,0 +1,151 @@
+	private void checkPreference(int displayedTitleResourceString, String sharedPreferenceTag) {
+		SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(InstrumentationRegistry.getTargetContext());
+
+		onData(PreferenceMatchers.withTitle(displayedTitleResourceString)).perform(click());
```
please reformat longer espresso calls like this to format:
onWhatever(.......)
     .doWhatever(....)

eg. 
onData(PreferenceMatchers.withTitle(displayedTitleResourceString))
      .perform(click());

84n4n4 commented:
test this please

84n4n4 commented:
please add the additional espresso tests to testsuites.AllEspressoTestsSuit.java to have them run with the pullrequest trigger...

84n4n4 commented:
please run espresso tests

84n4n4 commented:
please run espresso tests

84n4n4 commented:
ok, so to make the jenkins espresso PR trigger work, feature/NOLB would have to be rebased on develop. (since the PullRequestTriggerSuite is missing here...)

sjaindl commented:
OK, I have refactored the Espresso tests according to your last comments. 
Concerning the 2 tests in PhiroSettingsActivityTest.java, please see my comments (diffs outdated because I've changed the package). 
I can't rebase on develop now because we want don't want to have this branch ahead of the release branch for the upcoming Phiro release (I have to do that later to make PR-Trigger-Tests work).

cheidenreich commented on catroid/src/main/java/org/catrobat/catroid/ui/fragment/BaseProjectListFragment.java:
```diff
@@ -128,7 +127,7 @@ public int compare(ProjectData project1, ProjectData project2) {
 		});
 
-		projectAdapter = new ProjectListAdapter(getActivity(), R.layout.list_item, projectList);
+		projectAdapter = new ProjectListAdapter(context, R.layout.list_item, projectList);
```
`context` is null at this time, this causes the app to crash at older Android versions, changing it back to `getActivity()` works.

sjaindl replied:
Ok, I see. However, I can't rely on getActivity() here, because CopyPhiroProjectsTask in the ProjectListFragment of the Phiro-flavor causes the Fragment to be detached from it's activity. So, the context/activity must be remembered. 
So, I'm sure the issue is that, because we are not using the SupportFragmentManager, onAttach(Context) is simply not called. 
I think the best approach would be to override onAttach(Context) & onAttach(Activity) for API < 23, as described here: https://stackoverflow.com/questions/32077086/android-onattachcontext-not-called-for-api-23
